types/src/block/block_body.rs

```diff
@@ -1,10 +1,160 @@
 mod block_body_v1;
 mod block_body_v2;
-mod versioned_block_body;
 
 pub use block_body_v1::BlockBodyV1;
 pub use block_body_v2::BlockBodyV2;
-pub use versioned_block_body::VersionedBlockBody;
 
-/// The body portion of a block.
-pub type BlockBody = BlockBodyV2;
+use alloc::vec::Vec;
+use core::fmt::{self, Display, Formatter};
+
+#[cfg(feature = "datasize")]
+use datasize::DataSize;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    bytesrepr::{self, FromBytes, ToBytes, U8_SERIALIZED_LENGTH},
+    DeployHash,
+};
+
+const TAG_LENGTH: usize = U8_SERIALIZED_LENGTH;
+
+/// Tag for block body v1.
+pub const BLOCK_BODY_V1_TAG: u8 = 0;
+/// Tag for block body v2.
+pub const BLOCK_BODY_V2_TAG: u8 = 1;
+
+/// The versioned body portion of a block. It encapsulates different variants of the BlockBody
+/// struct.
+#[cfg_attr(feature = "datasize", derive(DataSize))]
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub enum BlockBody {
+    /// The legacy, initial version of the body portion of a block.
+    V1(BlockBodyV1),
+    /// The version 2 of the body portion of a block, which includes the
+    /// `past_finality_signatures`.
+    V2(BlockBodyV2),
+}
+
+impl BlockBody {
+    /// Retrieves the deploy hashes within the block.
+    pub fn deploy_hashes(&self) -> &Vec<DeployHash> {
+        match self {
+            BlockBody::V1(v1) => &v1.deploy_hashes,
+            BlockBody::V2(v2) => &v2.deploy_hashes,
+        }
+    }
+
+    /// Retrieves the transfer hashes within the block.
+    pub fn transfer_hashes(&self) -> &Vec<DeployHash> {
+        match self {
+            BlockBody::V1(v1) => &v1.transfer_hashes,
+            BlockBody::V2(v2) => &v2.transfer_hashes,
+        }
+    }
+
+    /// Returns deploy hashes of transactions in an order in which they were executed.
+    pub fn deploy_and_transfer_hashes(&self) -> impl Iterator<Item = &DeployHash> {
+        self.deploy_hashes()
+            .iter()
+            .chain(self.transfer_hashes().iter())
+    }
+}
+
+#[cfg(any(feature = "testing", test))]
+impl PartialEq for BlockBody {
+    fn eq(&self, other: &BlockBody) -> bool {
+        match (self, other) {
+            (BlockBody::V1(lhs), BlockBody::V1(rhs)) => lhs.eq(rhs),
+            (BlockBody::V2(lhs), BlockBody::V2(rhs)) => lhs.eq(rhs),
+            _ => {
+                panic!("BlockBody structs of different versions should not be compared")
+            }
+        }
+    }
+}
```

Comment:
Can't we just derive `PartialEq` and `Eq` rather than this impl with a panic?

Reply:
Fixed in [this commit](https://github.com/casper-network/casper-node/pull/4213/commits/14bcef4cbe729a6359e13189a20c18c4c0149da9) in the [follow-up PR](https://github.com/casper-network/casper-node/pull/4213).